pyomo/core/tests/unit/test_set.py

```diff
@@ -1768,6 +1769,24 @@ def x_init(m,i):
         self.assertEqual(i.x[1].domain, i.A*i.B)
         self.assertEqual(i.x[1], [])
 
+    @unittest.skipIf(not pandas_available, "pandas is not available")
+    def test_pandas_multiindex_set_init(self):
+        # Test that TuplizeValuesInitializer does not assume truthiness
+        # If it does, pandas will complain with the following error:
+        # ValueError: The truth value of a MultiIndex is ambiguous. 
+        # Use a.empty, a.bool(), a.item(), a.any() or a.all().
+        iterables = [['bar', 'baz', 'foo', 'qux'], ['one', 'two']]
+        pandas_index = pd.MultiIndex.from_product(
+            iterables, 
+            names=['first', 'second']
+        )
+
+        model = ConcreteModel()
+        model.a = Set(initialize=pandas_index,
+                      dimen=pandas_index.nlevels)
```

Comment:
Do you have to specify the dimen?  Is there a second test where you leave the dimen unspecified and Pyomo (hopefully correctly) infers that the set is 2D (and of course after creating the set, we check that dimen ends up 2)?

Reply:
Setting `dimen` explicitly is actually what triggers the call that leads to the ValueError but that's a reasonable test in and of itself. 

When not setting dimen explicitly in the example, the set is created appropriately but the following warning is raised which seems unexpected:
```
WARNING: Implicitly replacing the Component attribute a (type=<class
    'pyomo.core.base.set.OrderedSimpleSet'>) on block unknown with a new
    Component (type=<class 'pyomo.core.base.set.AbstractOrderedSimpleSet'>).
    This is usually indicative of a modelling error. To avoid this warning,
    use block.del_component() and block.add_component().
```

Any idea what the "modelling error" might be?

Here is the example:
```python
        iterables = [['bar', 'baz', 'foo', 'qux'], ['one', 'two']]
        pandas_index = pd.MultiIndex.from_product(
            iterables, 
            names=['first', 'second']
        )

        model = ConcreteModel()
        model.a = Set(initialize=pandas_index) # not specifying dimen
        # this will raise the warning but list(model.a) returns what we expect
```

Reply:
Yeah - you are using the same model for the two "halves" of the test.  The first part is creating `model.a`, then in this part you are creating another Set and attempting to also add it to the model using the `a` attribute.  The quick fix is to make the second `model.a = ` to just be `model.b = `.

Reply:
[Here](https://github.com/Pyomo/pyomo/blob/f607ebf301165caa8c78a3d5806fd5fcd6f946ec/pyomo/core/base/block.py#L565) is the line that gives the warning. Bit tricky to trace back without a stack trace (this doesn't seem like a big issue but flagging it anyways).

Reply:
Ahh 🤦‍♂️. Makes sense